### PR TITLE
Add information about the PreprocessString function

### PR DESCRIPTION
**Pull request by EleventhNov**

### CODE REVIEW

File: headers/types/common/common.h [hunk 1]
```diff
@@ -36,6 +36,26 @@ struct file_stream {
 };
 ASSERT_SIZE(struct file_stream, 72);
 
+// These flags are shared with the function to display text inside message boxes
+// So they might need a rename once more information is found
+struct preprocessor_flags {
+    uint16_t unknown : 13;
+    bool show_speaker : 1;
+    uint16_t unknown : 18;
+}
```
Comment (UsernameFodder) [inline headers/types/common/common.h:45]: Need a semicolon here

File: headers/types/common/common.h [hunk 2]
```diff
@@ -36,6 +36,26 @@ struct file_stream {
...
+
+// Represents arguments that might be passed to the PreprocessString function
+struct preprocessor_args {
+    uint32_t flag_vals[4]; // These are usually IDs with additional flags attached to them
+    uint32_t id_vals[5];
+    int32_t number_vals[5]; 
+    char* strings[5];
+    uint32_t speaker_id; // An optional argument that is used to insert the name of a Pokémon
+    // When they're talking through a message box. It requires it's respective flag to be on
+}
```
Comment (UsernameFodder) [inline headers/types/common/common.h:56]: Semicolon here

File: symbols/arm9.yml [hunk 3]
```diff
@@ -414,12 +414,28 @@ arm9:
...
         
         return: language ID
+    - name: PreprocessString
+      address:
+        NA: 0x20223F0
+        EU: 0x20225EC
+      description: |-
+        An enhanced sprintf, which recognizes certain tags and replaces them with appropiate game values.
+        This function can also be used to simply insert values passed within the preprocessor args
+
```
Comment (UsernameFodder) [inline symbols/arm9.yml:424]: Kind of annoying, but with YAML you need to have spaces at the start of empty lines to match the indentation of the paragraphs. Same thing with the other blank line.

**Review by UsernameFodder — CHANGES_REQUESTED** [review]
Looks good, just a couple syntactical/formatting issues to fix.

**Comment by EleventhNov** [comment]
Oh boy that's embarrasing lol, I really need to get a syntax checker... Hopefully this fixes it tho!

**Comment by UsernameFodder** [comment]
> Oh boy that's embarrasing lol, I really need to get a syntax checker... Hopefully this fixes it tho!

It'll be a bit more annoying on this PR since it's your first and GitHub asks me for approval before the automatic checks will run (apparently it was to stop cryptomining abuse! :P). On subsequent PRs you'll be able to use the pipeline results as a syntax checker without needing to worry about setting anything up yourself.

File: headers/types/common/common.h [hunk 4]
```diff
@@ -36,6 +36,26 @@ struct file_stream {
...
+// Represents arguments that might be passed to the PreprocessString function
+struct preprocessor_args {
+    uint32_t flag_vals[4]; // These are usually IDs with additional flags attached to them
+    uint32_t id_vals[5];
+    int32_t number_vals[5]; 
+    char* strings[5];
+    uint32_t speaker_id; // An optional argument that is used to insert the name of a Pokémon
+    // When they're talking through a message box. It requires it's respective flag to be on
+};
+ASSERT_SIZE(preprocessor_args, 80);
```
Comment (UsernameFodder) [inline headers/types/common/common.h:57]: One more compile issue, use `struct preprocessor_args` here.

File: headers/types/common/common.h [hunk 5]
```diff
@@ -36,6 +36,26 @@ struct file_stream {
...
+ASSERT_SIZE(struct preprocessor_flags, 4);
+
+// Represents arguments that might be passed to the PreprocessString function
+struct preprocessor_args {
+    uint32_t flag_vals[4]; // These are usually IDs with additional flags attached to them
+    uint32_t id_vals[5];
+    int32_t number_vals[5]; 
+    char* strings[5];
+    uint32_t speaker_id; // An optional argument that is used to insert the name of a Pokémon
+    // When they're talking through a message box. It requires it's respective flag to be on
```
Comment (UsernameFodder) [inline headers/types/common/common.h:55]: Does this comment refer to the above `speaker_id`? If so, could you move the comments to the line above the field? This is the style I've been using:
```c
    // An optional argument that is used to insert the name of a Pokémon.
    // When they're talking through a message box. It requires its respective flag to be on
    uint32_t speaker_id;
```

**Review by UsernameFodder — CHANGES_REQUESTED** [review]

File: headers/functions/arm9.h [hunk 6]
```diff
@@ -48,6 +48,8 @@ enum move_category GetMoveCategory(enum move_id move_id);
 void HandleSir0Translation(uint8_t** dst, uint8_t* src);
 void HandleSir0TranslationVeneer(uint8_t** dst, uint8_t* src);
 int GetLanguage(void);
+int PreprocessString(char* output, int output_size, char* format, 
+                     struct preprocessor_flags flags, struct preprocessor_args* args);
```
Comment (UsernameFodder) [inline headers/functions/arm9.h:52]: [Formatter caught](https://github.com/UsernameFodder/pmdsky-debug/runs/5298100158?check_suite_focus=true) that there's enough room to put `struct preprocessor_flags flags` on the previous line.

File: headers/types/common/common.h [hunk 7]
```diff
@@ -36,6 +36,26 @@ struct file_stream {
...
+    bool show_speaker : 1;
+    uint32_t unknown18 : 18;
+};
+ASSERT_SIZE(struct preprocessor_flags, 4);
+
+// Represents arguments that might be passed to the PreprocessString function
+struct preprocessor_args {
+    uint32_t flag_vals[4]; // These are usually IDs with additional flags attached to them
+    uint32_t id_vals[5];
+    int32_t number_vals[5];
```
Comment (UsernameFodder) [inline headers/types/common/common.h:52]: [Formatter caught]([Formatter caught](https://github.com/UsernameFodder/pmdsky-debug/runs/5298100158?check_suite_focus=true) that there's enough room to put `struct preprocessor_flags flags` on the previous line.) that there's an extra space at the end of this line :o.

**Review by UsernameFodder — CHANGES_REQUESTED** [review]

File: headers/types/common/common.h [hunk 8]
```diff
@@ -104,4 +124,4 @@ ASSERT_SIZE(struct move_data_table, 14534);
 // TODO: Add more data file structures, as convenient or needed, especially if the load address
 // or pointers to the load address are known.
 
-#endif
```
Comment (UsernameFodder) [inline headers/types/common/common.h:107]: Err, let's keep the empty line at the end of the file as is. It makes things look a little nicer in the GitHub viewer.

**Review by UsernameFodder — CHANGES_REQUESTED** [review]

**Comment by EleventhNov** [comment]
Cryptomining abuse huh? That's something I didn't expect to hear here today haha, Anyway I apologize for the clunkiness, Hopefully this fixes it for good

**Review by UsernameFodder — APPROVED** [review]
Looks good now. Thanks for contributing!

**Pull request merged**